nixos/modules/services/printing/cups-pdf.nix

```diff
@@ -0,0 +1,205 @@
+{ config, lib, pkgs, ... }:
+
+let
+
+  inherit (lib.attrsets) filterAttrs getAttr mapAttrs mapAttrsToList;
+  inherit (lib.meta) getExe;
+  inherit (lib.modules) mkAfter mkDefault mkIf;
+  inherit (lib.options) literalExpression mkEnableOption mkOption;
+  inherit (lib.strings) concatStrings;
+  inherit (lib.trivial) pipe;
+  inherit (lib.types) attrsOf int lines nullOr oneOf package path singleLineStr str submodule;
+
+  # cups calls its backends as user `lp` (which is good!),
+  # but cups-pdf wants to be called as `root`,
+  # so it can change ownership of files.
+  # We add a suid wrapper and a wrapper script
+  # to trick cups into calling the suid wrapper.
+
+  # wrapper script that redirects calls to the suid wrapper
+  wrapper = pkgs.writeTextFile {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapper.sh";
+    executable = true;
+    destination = "/lib/cups/backend/cups-pdf";
+    checkPhase = ''
+      ${pkgs.stdenv.shellDryRun} "$target"
+      ${getExe pkgs.shellcheck} "$target"
+    '';
+    text = ''
+      #! ${pkgs.runtimeShell}
+      exec "${config.security.wrapperDir}/cups-pdf" "$@"
+    '';
+  };
+
+  # wrapped cups-pdf package that uses the suid wrapper
+  wrapped = pkgs.buildEnv {
+    name = "${pkgs.cups-pdf-to-pdf.name}-wrapped";
+    # using wrapper as first path ensures the wrapper is used
+    paths = [ wrapper pkgs.cups-pdf-to-pdf ];
+    ignoreCollisions = true;
+  };
+
+  instanceSettings = name: {
+    freeformType = nullOr (oneOf [ int str path package ]);
+    # override defaults:
+    # inject instance name into paths,
+    # also avoid conflicts between user names and special dirs
+    options.Out = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/users/\${USER}";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/users/\${USER}";
+      example = "\${HOME}/cups-pdf";
+      description = "output directory";
+    };
+    options.AnonDirName = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/anonymous";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp";
+      description = "path for anonymously created PDF files";
+    };
+    options.Spool = mkOption {
+      type = nullOr singleLineStr;
+      default = "/var/spool/cups-pdf-${name}/spool";
+      defaultText = "/var/spool/cups-pdf-{instance-name}/anonymous";
+      example = "/tmp/cups-pdf";
+      description = "spool directory";
+    };
+    options.Anonuser = mkOption {
+      type = nullOr singleLineStr;
+      default = "";
+      example = "root";
+      description = ''
+        User for anonymous PDF creation.
+        An empty value disables this feature.
+        Avoid the default value <literal>nobody</literal>
+        as it might cause security issues.
```

Comment:
We should assert that.

Reply:
I'm not sure what the best solution is here.  I now set the default to `root` which might annoy anonymous clients but not pose any security risk.  To avoid the `nobody` account being used, I removed the `nullOr` type.  This means one should not be able to assign print jobs to the `nobody` account unless one directly sets this option to `nobody`.

The description also no longer mentiones `nobody`.  Is it acceptable to allow the user to explicitely use the `nobody` group if s/he really wants to?